### PR TITLE
Refactor llama3 demo to the new generator API

### PR DESCRIPTION
**Pull request by mtairum**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/16752

### What's changed
- New Llama3 demo now uses the generator API
- Improved prefill performance. E.g. Llama3-70B now at 182ms prefill time
- Improved profiling in the demo
- Removed old text demo and updated CI accordingly
- Cleaned up the prompt input files and added missing ones.
- New benchmark profiling for superset: now includes TTFT, and full decode perf for 4096 iteration (for plotting).
- Add llama3 demo custom input support: you can now override any settings for easier testing.
- Updated PERF.md with the latest numbers.

### Checklist
- [x] [All-post-commit](https://github.com/tenstorrent/tt-metal/actions/runs/13058153825)
- [x] [T3K](https://github.com/tenstorrent/tt-metal/actions/runs/13058160868)
- [x] [Single](https://github.com/tenstorrent/tt-metal/actions/runs/13070314473)
- [x] [TG-demo](https://github.com/tenstorrent/tt-metal/actions/runs/13058170374)
- [x] [T3K accuracy](https://github.com/tenstorrent/tt-metal/actions/runs/13069483167/job/36468085102)

### CODE REVIEW

File: models/demos/llama3/tt/llama_model.py [hunk 1]
```diff
@@ -242,23 +242,22 @@ def process_output_prefill(self, tt_out, last_token_idx):
...
-                    cluster_axis=0,
-                    mesh_device=self.mesh_device,
-                    topology=ttnn.Topology.Linear,
-                )
-            else:
-                tt_out = ttnn.all_gather(tt_out, dim=3, num_links=1, topology=ttnn.Topology.Linear)
-        tt_out = ttnn.untilize(tt_out, use_multicore=True)
+        if argmax_on_device:
+            tt_out = ttnn.to_torch(
+                # tt_out.cpu(blocking=True, cq_id=1),
```
Comment (skhorasganiTT) [inline models/demos/llama3/tt/llama_model.py:251]: Can this comment be removed?

File: models/demos/llama3/tt/generator.py [hunk 2]
```diff
@@ -214,6 +218,26 @@ def _decode_forward_no_trace_text(
...
+                )
+            else:
+                tt_logits = ttnn.all_gather(tt_logits, dim=3, num_links=1, topology=ttnn.Topology.Linear)
+        tt_logits = ttnn.untilize(tt_logits, use_multicore=True)
+
+        if argmax_on_device:
+            tt_logits = ttnn.argmax(  # TODO Add multicore support to batch > 1
+                tt_logits, dim=3, use_multicore=False if tokens.shape[0] > 1 else True  # ,output_tensor=tokens
+            )
+
```
Comment (skhorasganiTT) [inline models/demos/llama3/tt/generator.py:240]: Can this be moved inside `ttnn_decode_forward` so that it's not replicated in `_capture_trace_text`?

File: models/demos/llama3/tt/generator.py [hunk 3]
```diff
@@ -503,6 +549,20 @@ def _decode_forward_no_trace(
...
+                    dim=3,
+                    num_links=2,
+                    cluster_axis=0,
+                    mesh_device=self.model.mesh_device,
+                    topology=ttnn.Topology.Linear,
+                )
+            else:
+                tt_logits = ttnn.all_gather(tt_logits, dim=3, num_links=1, topology=ttnn.Topology.Linear)
+        tt_logits = ttnn.untilize(tt_logits, use_multicore=True)
+
```
Comment (skhorasganiTT) [inline models/demos/llama3/tt/generator.py:565]: Same comment as before except for the vision model, can this be moved inside ttnn_decode_forward so it doesn't need to be copied in `_capture_trace` (currently missing there)

File: models/demos/llama3/tt/generator.py [hunk 4]
```diff
@@ -692,7 +752,7 @@ def _decode_forward_trace(
             ),
         )
 
-        ttnn.execute_trace(self.mesh_device, trace_id, cq_id=0, blocking=False)
+        ttnn.execute_trace(self.mesh_device, trace_id, cq_id=0, blocking=True)
```
Comment (skhorasganiTT) [inline models/demos/llama3/tt/generator.py:755]: Why was this modified to blocking True?

File: models/demos/llama3/tt/generator.py [hunk 5]
```diff
@@ -841,8 +901,8 @@ def chat_completion(
         top_p: float = 0.9,
         max_gen_len=None,
     ):
-        if max_gen_len is None or max_gen_len == 0 or max_gen_len >= self.model.configuration.max_seq_len:
-            max_gen_len = self.model.configuration.max_seq_len - 1
+        # if max_gen_len is None or max_gen_len == 0 or max_gen_len >= self.model.configuration.max_seq_len:
+        # max_gen_len = self.model.configuration.max_seq_len - 1
```
Comment (skhorasganiTT) [inline models/demos/llama3/tt/generator.py:905]: Was the meant to be commented?

File: models/demos/llama3/tt/generator.py (hunk 3 above)
Comment (skhorasganiTT) [reply models/demos/llama3/tt/generator.py:565]: In addition, it might be good to put it in a util function that is used in both the `ttnn_decode_forward` of the text and vision models so we don't get divergence there

**Review by skhorasganiTT — COMMENTED** [review]
Reviewed all the files except simple_text_demo.py for now since that one is still being cleaned up

File: models/demos/llama3/tt/llama_model.py (hunk 1 above)
Comment (mtairum) [reply models/demos/llama3/tt/llama_model.py:251]: I will leave this one since in the past we had issues with this call not being blocked. The comment will serve as a reminder if we end up seeing hangs again in the future.

File: models/demos/llama3/tt/generator.py (hunk 3 above)
Comment (mtairum) [reply models/demos/llama3/tt/generator.py:565]: What do you mean by this util function?

Comment (skhorasganiTT) [reply models/demos/llama3/tt/generator.py]: They don't have the same function, there is `llama_model.py::TtTransformer::ttnn_decode_forward` for the text model and `llama_vision_model.py::CrossAttentionTransformer::ttnn_decode_forward` for vision. So this code will need to be in both probably, which is why I was suggesting to make a function for it and call it in both places.

Comment (skhorasganiTT) [reply models/demos/llama3/tt/generator.py]: But I just remembered that the vision model also has it's own `process_output_decode` and it was never using this code so nevermind

File: models/demos/llama3/tt/generator.py [hunk 6]
```diff
@@ -449,8 +459,8 @@ def decode_forward(
         else:
             return tt_logits
 
-    def read_decode_output(self, tt_logits, unpadded_batch):
-        logits = self.model.process_output_decode(tt_logits, B=unpadded_batch, S=1)
+    def read_decode_output(self, tt_logits, unpadded_batch, argmax_on_device=False):
+        logits = self.model.process_output_decode(tt_logits, B=unpadded_batch, S=1, argmax_on_device=argmax_on_device)
```
Comment (skhorasganiTT) [inline models/demos/llama3/tt/generator.py:463]: `llama_vision_model.py::ttnn_decode_forward` and `llama_vision_model.py::process_output_decode` are currently missing the argmax_on_device flag and will fail

**Review by skhorasganiTT — APPROVED** [review]

**Review by tt-rkim — APPROVED** [review]

**Comment by tt-rkim** [comment]
Did you run nightly ttnn single card, t3k demos, tg demos?

**Comment by mtairum** [comment]
@tt-rkim In the process of fixing the CI issues. Will only merge after they pass 👍

**Comment by mtairum** [comment]
This PR is now feature complete, with a bunch of new benchmark functionality and QoL updates to the demo to better measure what the customer team throws at us. 

Waiting for CI to pass then it's good to go.

**Review by yieldthought — APPROVED** [review]
_Clean_

File: models/demos/llama3/scripts/op_perf_results.py
[the remark this answers is not in the document]
Comment (mtairum) [reply models/demos/llama3/scripts/op_perf_results.py:1]: Re-added it 👍

File: models/demos/llama3/tt/llama_common.py [hunk 7]
```diff
@@ -112,8 +189,8 @@ def gather_cos_sin(position_ids, cos, sin):
     return cos, sin
 
 
-def get_prefill_rot_mat(head_dim, max_seq_len, mesh_device, seq_len, scale_factor, start_pos=0):
-    cos, sin = precompute_freqs(head_dim, max_seq_len * 2, scale_factor=scale_factor)
+def get_prefill_rot_mat(head_dim, mesh_device, seq_len, scale_factor, start_pos=0):
+    cos, sin = precompute_freqs(head_dim, seq_len * 2, scale_factor=scale_factor)
```
Comment (cglagovichTT) [inline models/demos/llama3/tt/llama_common.py:193]: is this correct? Why is it now taking seq_len rather than max_seq_len?

Comment (mtairum) [reply models/demos/llama3/tt/llama_common.py:193]: `Max_seq_len` is ideally specified to the max length supported by the model. So typically 128k.

`seq_len` in this function's context is the prefill length. This avoids running the `precompute_freqs` for 2*128k, if the prefll seqlen is smaller. For larger models these savings are significant.

For Llama70B, for smaller prompts (<16K) we can save up 20ms of prefill time (184ms vs. 204ms).

For decode we run a separate function for the rot mats.

Comment (cglagovichTT) [reply models/demos/llama3/tt/llama_common.py:193]: doesn't that mean that the freqs could be different between prefill and decode? Decode freqs are based on `max_seq_len`, prefill are based on prefill `seqlen`?

Comment (cglagovichTT) [reply models/demos/llama3/tt/llama_common.py:193]: We already have cos/sin matrices in DRAM for decode. We should solve the e2e performance issue by having prefill slice those matrices to get the cos/sin it needs

**Comment by mtairum** [comment]
All CI are passing. 
But due to the higher priority of Qwen2.5 and Deepseek distilled models, the merge of this branch will wait for that to hit main to avoid delays.

Godspeed @yieldthought 🫡

**Comment by mtairum** [comment]
After other commits got merged in, it's time for this PR to go. 

Fix conflicts, and re-did all tests:
- [ ] [All-post-commit](https://github.com/tenstorrent/tt-metal/actions/runs/13433726398)
- [x] [Single-device](https://github.com/tenstorrent/tt-metal/actions/runs/13420815135)
- [x] [T3K](https://github.com/tenstorrent/tt-metal/actions/runs/13420819678)
  - The demo pipeline is currently hanging in CI. Passing locally.
- [ ] [T3K-perplexity](https://github.com/tenstorrent/tt-metal/actions/runs/13433796202)
  - The perplexity pipeline is currently hanging in CI. Passing locally.
